tokio/src/sync/notify.rs

```diff
@@ -0,0 +1,496 @@
+                                        *state = Done;
+                                        continue 'outer;
+                                    }
+                                    Err(actual) => {
+                                        curr = actual;
+                                        continue;
+                                    }
+                                }
+                            }
+                            _ => unreachable!(),
+                        }
+
+                        // The state has been transitioned to waiting
+                        break;
+                    }
+
+                    // Safety: called while locked.
+                    unsafe {
+                        (*waiter.as_mut().get()).waker = Some(cx.waker().clone());
+
+                        // Insert the waiter into the linked list
+                        waiters.push_front(waiter.as_mut());
+                    }
+
+                    *state = Waiting;
+                }
+                Waiting => {
+                    let mut waiters = notify.waiters.lock().unwrap();
+
+                    {
+                        // Safety: called while locked
+                        let w = unsafe { &mut *waiter.as_mut().get() };
+
+                        if w.notified {
+                            w.waker = None;
+                            w.notified = false;
+
+                            // Remove the entry from the linked list. The entry
+                            // may only be in the linked list while the state is
+                            // `Waiting`.
+                            //
+                            // Safety: called while the lock is held
+                            unsafe {
+                                waiters.remove(waiter.as_mut());
+                            }
+
+                            *state = Done;
+                        } else {
+                            if !w.waker.as_ref().unwrap().will_wake(cx.waker()) {
+                                w.waker = Some(cx.waker().clone());
+                            }
+
+                            return Poll::Pending;
+                        }
+                    }
+                }
+                Done => {
+                    return Poll::Ready(());
+                }
+            }
+        }
+    }
+}
+
+impl Drop for RecvFuture<'_> {
+    fn drop(&mut self) {
+        use RecvState::*;
+
+        // Safety: The type only transitions to a "Waiting" state when pinned.
+        let (notify, state, mut waiter) = unsafe { Pin::new_unchecked(self).project() };
+
+        // This is where we ensure safety. The `RecvFuture` is being dropped,
+        // which means we must ensure that the waiter entry is no longer stored
+        // in the linked list.
+        if let Waiting = *state {
+            let mut notify_state = WAITING;
+            let mut waiters = notify.waiters.lock().unwrap();
+
+            // remove the entry from the list
```

Comment:
I'm surprised we don't have to re-check that the state is still `WAITING` here. What if just before we took the lock, a notify happened which notified us, removed us from the list, and set the state to `EMPTY`?

Reply:
It took me a moment to remember, but I'm pretty sure it is correct. I added comments explaining.